lib/GHCup/Prelude/MegaParsec.hs

```diff
@@ -137,3 +137,26 @@ isSpace :: Char -> Bool
 isSpace c = (c == ' ') || ('\t' <= c && c <= '\r')
 {-# INLINE isSpace #-}
 
+-- Obtain the version from the link or shim path
+-- ../ghc/<ver>/bin/ghc
+-- ../ghc/<ver>/bin/ghc-<ver>
+ghcVersionFromPath :: MP.Parsec Void Text GHCTargetVersion
+ghcVersionFromPath =
+  (do
+     beforeBin <- parseUntil1 binDir
+     rest  <- MP.getInput
+     MP.setInput beforeBin
+     _ <- parseTillLastPathSep
+     x <- ghcTargetVerP
+     MP.setInput rest
+     pure x
+   )
+  <* MP.some pathSep
+  <* MP.takeRest
+  <* MP.eof
+  where
+     binDir = MP.some pathSep <* MP.chunk "bin" *> MP.some pathSep <* MP.chunk "ghc"
+     parseTillLastPathSep = MP.manyTill MP.anySingle (MP.try $ do
+       _ <- MP.some pathSep
+       _ <- MP.lookAhead (MP.many MP.anySingle)
+       MP.notFollowedBy (MP.manyTill MP.anySingle (MP.some pathSep)))
```

Comment:
I don't understand this.

Especially since `MP.lookAhead (MP.many MP.anySingle)` always parses until end of input.

Reply:
I think we can just do:

```hs
     parseTillLastPathSep = (MP.try (parseUntil1 pathSep *> many pathSep) *> p) <|> many MP.anySingle
```

Reply:
```hs
ghcVersionFromPath :: MP.Parsec Void Text GHCTargetVersion
ghcVersionFromPath =
  do
     beforeBin <- parseUntil1 binDir <* MP.some pathSep
     MP.setInput beforeBin
     _ <- parseTillLastPathSep
     ghcTargetVerP
  where
     binDir = MP.some pathSep <* MP.chunk "bin" *> MP.some pathSep <* MP.chunk "ghc"
     parseTillLastPathSep = (MP.try (parseUntil1 pathSep *> MP.some pathSep) *> parseTillLastPathSep) <|> pure ""
```

This is much cleaner.

The `pure ""` in `parseTillLastPathSep` is important so we always break out of the parser in the first failure (innermost recursion)... if we would parser the ghc version right there and it's a borked ghc version, it would try to re-parse for every filepath component

Reply:
Ok, yes, recursive is much cleaner.